### ISSUE TITLE
Propagate shard specs through composite ops.

### ISSUE THREAD

**Issue by AleksKnezevic** [tenstorrent/tt-xla#1838]
(acolicTT) Summary of a problem:

Running: tests/runner/test_models.py::test_all_models[gemma/pytorch-google/gemma-1.1-7b-it-tensor_parallel-full-inference] on a n300.

When stable HLO composite ops are created, they look something like this (this is pytorch gelu composite op created using pytorch's stableHLOCompositeBuilder):
```
%150 = stablehlo.composite "tenstorrent.gelu_tanh" %149 {composite_attributes = {approximate = "tanh"}, decomposition = @tenstorrent.gelu_tanh.impl_17} : (tensor<1x512x24576xbf16>) -> tensor<1x512x24576xbf16> loc(#loc398)
...
func.func private @tenstorrent.gelu_tanh.impl_17(%arg0: tensor<1x512x24576xbf16> {ttcore.argument_type = #ttcore.argument_type<input>} loc("custom-call.1668")) -> tensor<1x512x24576xbf16> {
...
%0 = stablehlo.broadcast_in_dim %cst_3, dims = [] : (tensor<bf16>) -> tensor<1x512x24576xbf16> loc(#loc)
%1 = stablehlo.broadcast_in_dim %cst_2, dims = [] : (tensor<bf16>) -> tensor<1x512x24576xbf16> loc(#loc)
...
return %12 : tensor<1x512x24576xbf16> loc(#loc)
} loc(#loc)
```

When shardy shards this composite op, it changes only stablehlo.composite op's input and output type from tensor<1x512x24576xbf16> to tensor<1x512x12288xbf16>.

On the other hand, decomposition (func.func private @tenstorrent.gelu_tanh.impl_17) is not handled by shardy and it's types stays that same. I checked shardy repo, and there is no support for composite ops: `shardy/dialect/sdy/transforms/propagation/op_sharding_rule_registry.cc`
It is handled with some generic code that just propagates operation attiributes.

After shardy pass, stablehlo.composite has different type:
```
%150 = stablehlo.composite "tenstorrent.gelu_tanh" %149 {composite_attributes = {approximate = "tanh"}, decomposition = @tenstorrent.gelu_tanh.impl_17} : (tensor<1x512x12288xbf16>) -> tensor<1x512x12288xbf16> loc(#loc398)
...
```
but decomposition stay the same:
```
func.func private @tenstorrent.gelu_tanh.impl_17(%arg0: tensor<1x512x24576xbf16> {ttcore.argument_type = #ttcore.argument_type<input>} loc("custom-call.1668")) -> tensor<1x512x24576xbf16> {
...
%0 = stablehlo.broadcast_in_dim %cst_3, dims = [] : (tensor<bf16>) -> tensor<1x512x24576xbf16> loc(#loc)
%1 = stablehlo.broadcast_in_dim %cst_2, dims = [] : (tensor<bf16>) -> tensor<1x512x24576xbf16> loc(#loc)
...
return %12 : tensor<1x512x24576xbf16> loc(#loc)
} loc(#loc)
```
There is a stable HLO pass that checks composite ops in stable HLO repo in `stablehlo/dialect/TypeInference.cpp`, and one of the checks is to match composite op type and it's decomposition types. In our case it fails, since shardy only changed composite op, but not it's decomposition.

So we are getting an error: `loc("custom-call.1685"): error: operand at index 0 has type 'tensor<1x512x12288xbf16>', but decomposition has type 'tensor<1x512x24576xbf16>'`

I asked on a openXLA discord channel whether they plan to support composite ops with shardy. Now, I am waiting for the response.

**Comment by nvukobratTT** [comment]
This is a blocker to enable composite ops through PyTorch. @AleksKnezevic, can we prioritize this issue?

**Comment by nvukobratTT** [comment]
Even if we determine that we need to add support for each composite op (i.e., not use shlo impl from the composite for sharding), there is still a value from the MLIR perspective as we're simplifying fusing. Therefore, let's discuss what is the best course of action here, and outline when it's more appropriate to use composites, and when we should fallback to custom calls.

**Comment by AleksKnezevic** [comment]
Discussed offline, key points:

- Cannot have custom sharding logic for each composite op
- We need a solution that's either all local (i.e. passes that we call) or can be upstreamed, current solution cannot be upstreamed
- @sshonTT will try to flatten the graph in order to propagate shard specs and re-capture composite op patterns after propagation
  - Any CCLs inserted within composite ops will break the composite and we'll fall back to the base ops. This is fine because a softmax with all_gather in the middle is inherently different than a single chip softmax. 
  - CCLs outside of composite ops cannot break the composite op. 
- Custom SHLO ops (non composite) will need to have their own sharding propagation rules.

**Comment by sshonTT** [comment]
[here](https://github.com/tenstorrent/tt-mlir/tree/sshon/composite-op-flatten-sharding-reoutline)'s my implementation for composite-op flatten → propagation → re-outline.

**Comment by sshonTT** [comment]
Merged the PR.

**Comment by nvukobratTT** [comment]
> Merged the PR.

Nice! 

FYI @mstojkovicTT, this should unblock composite ops support through Torch. @sshonTT, is the description is this PR still relative? I.e., high-level explanation of the feature is correct?
- https://github.com/tenstorrent/tt-mlir/pull/5729#issue-3593101827
